BFT Block Puller: censorship monitor

tock-ibm opened the pull request:
#### Type of change
- New feature

#### Description

BFT Block Puller: censorship monitor

BFT header receivers that pull headers from the orderers, and keep the time and number of the last one.
The header receivers verify each block as it arrives.
The header receivers will receive (or ask for) full config blocks - in a later commit.
The header receivers will maintain their own private block verifier (bundle) - in a later commit.

A BFT censorship monitor which periodically compares the progress of headers relative to blocks.
The monitor triggers an alert if a header is ahead of the block stream for more than a certain time.

The BFTDeliverer is only a skeleton.

#### Related issues
#4306

yacovm commented on internal/pkg/peer/blocksprovider/bft_censorship_monitor.go:
```diff
@@ -0,0 +1,277 @@
+)
+
+type BlockProgressReporter interface {
+	BlockProgress() (uint64, time.Time)
```
which struct implements this?

tock-ibm replied:
This will be implemented by the BFTDeliverer, which keeps track of block arrival.

yacovm commented on internal/pkg/peer/blocksprovider/bft_header_receiver.go:
```diff
@@ -0,0 +1,164 @@
+}
+
+// CloseSend closes the client connection
+func (hr *bftHeaderReceiver) CloseSend() error {
```
should we not wait until `DeliverHeaders()` exits?

tock-ibm replied:
Not really. Actually `DeliverHeaders` calls this one as well when it exits. 
This is used to send a signal to the orderer to stop sending headers and close the connection.
`DeliverHeaders` will pickup the error from the Recv() and exit.

yacovm commented on internal/pkg/peer/blocksprovider/bft_censorship_monitor.go:
```diff
@@ -0,0 +1,277 @@
+			if m.detectBlockCensorship() {
+				m.errorCh <- &errCensorship{message: fmt.Sprintf("block censorship detected, endpoint: %s", m.fetchSources[m.blockSourceIndex])}
+				m.logger.Debug("Block censorship detected, stopping to monitor progress")
+				m.Stop()
```
I don't understand. Who relaunches this function if we exit?

tock-ibm replied:
This method should be run using a dedicated goroutine.

yacovm commented on internal/pkg/peer/blocksprovider/bft_censorship_monitor.go:
```diff
@@ -0,0 +1,277 @@
+		return false
+	}
+
+	sort.Slice(ahead, func(i, j int) bool {
```
why do you sort this? You can just iterate over it and find the max/min no? 

I know the computation here is negligible, but this is just awkward to sort in order to get the minimum / maximum.

tock-ibm replied:
done.

yacovm commented on internal/pkg/peer/blocksprovider/bft_header_receiver.go:
```diff
@@ -0,0 +1,164 @@
+	}
+
+	if !hr.lastHeaderVerified {
+		hr.lastHeaderVerified = true
```
you are setting this to true prematurely, no? what if it's not verified in the end? don't you need to set it to false then?

tock-ibm replied:
when we receive a header we do 
```
// do not verify, just save for later, in case the block-receiver is suspected of censorship
			hr.logger.Debugf("[%s][%s] Saving block with header & metadata, blockNum = [%d], block = [%v]", hr.chainID, hr.endpoint, blockNum, t.Block)
			hr.mutex.Lock()
			hr.lastHeader = t.Block
			hr.lastHeaderTime = time.Now()
			hr.lastHeaderVerified = false
			hr.lastHeaderOK = false
			hr.mutex.Unlock()
```
L84

yacovm commented on internal/pkg/peer/blocksprovider/bft_header_receiver.go:
```diff
@@ -0,0 +1,164 @@
+		hr.lastHeaderVerified = true
+		hr.lastHeaderOK = true
+
+		err := hr.msgCryptoVerifier.VerifyBlockAttestation(hr.chainID, hr.lastHeader)
```
Why do you lazily verify the attestation when you probe for it and not when you ingest it? What's the point of it? You anyway process through all block headers here, no?

tock-ibm replied:
It is lazy because we test the front of the header receivers array every T ms.

If T is larger compared to the block rate (say T=100ms, and rate=1k block/sec), we save a lot of unnecessary verifications (100x saving, resp).

If T is small compared to the block rate (say T=100ms, and rate = 1 block/sec), we use the verification caching feature, i.e. we only verify a header once and remember what the result was (10x saving, resp.).

Sig. verification is relatively costly, so we want to avoid unnecessary verifications.

yacovm commented on internal/pkg/peer/blocksprovider/bft_censorship_monitor.go:
```diff
@@ -0,0 +1,277 @@
+		}
+	}
+
+	if len(ahead) == 0 {
```
I don't think this is correct. 

We can have a few headers ahead of us, can we not? 

We should simply not have headers ahead of us with timestamps that are too in the past. 
What we should really look for, is the timestamp of when we received our last block and when was the last block header received from the header pullers, don't you agree?

tock-ibm replied:
When we have a header ahead of us, we raise a suspicion and record the time. We then wait for a censorship-timeout before we declare the suspicion as a censorship event.

If a block arrives before the timeout expires we remove the suspicion.

yacovm commented on internal/pkg/peer/blocksprovider/bft_censorship_monitor.go:
```diff
@@ -0,0 +1,277 @@
+	headerVerifier   BlockVerifier
+	requester        *DeliveryRequester
+	fetchSources     []*orderers.Endpoint
+	blockSourceIndex int
```
the source index and sources are never updated.

tock-ibm replied:
True, it is only set in the constructor method. The current plan is to stop the monitor and recreate it when ever we 
- refresh endpoints
- detect censorship (i.e. change the fetch index)

Since these are rare events, I prefer simplicity.

tock-ibm committed:
BFT Block Puller: censorship monitor

BFT header receivers that pull headers from the orderers, and keep the time and number of the last one.
The header receivers verify each block as it arrives.
The header receivers will receive (or ask for) full config blocks - in a later commit.
The header receivers will maintain their own private block verifier (bundle) - in a later commit.

A BFT censorship monitor which periodically compares the progress of headers relative to blocks.
The monitor triggers an alert if a header is ahead of the block stream for more than a certain time.

The BFTDeliverer is only a skeleton

Signed-off-by: Yoav Tock <tock@il.ibm.com>
Change-Id: I5180a98640073b87effb478fd86b6fa7d4df5dee
